src/tallies/filter_collision.cpp

```diff
@@ -0,0 +1,121 @@
+#include "openmc/tallies/filter_collision.h"
+
+#include <fmt/core.h>
+
+#include "openmc/capi.h"
+#include "openmc/constants.h" // For F90_NONE
+#include "openmc/mgxs_interface.h"
+#include "openmc/search.h"
+#include "openmc/settings.h"
+#include "openmc/xml_interface.h"
+
+namespace openmc {
+
+//==============================================================================
+// CollisionFilter implementation
+//==============================================================================
+
+void CollisionFilter::from_xml(pugi::xml_node node)
+{
+  auto bins = get_node_array<int>(node, "bins");
+  this->set_bins(bins);
+}
+
+void CollisionFilter::set_bins(gsl::span<const int> bins)
+{
+  // Clear existing bins
+  bins_.clear();
+  bins_.reserve(bins.size());
+
+  // Copy bins, ensuring they are valid
+  for (gsl::index i = 0; i < bins.size(); ++i) {
+    if (i > 0 && bins[i] <= bins[i - 1]) {
```

Comment:
Two things:

1) This is a possible access out-of-bounds. You check `i>0 && bins[i] <= bins[i-1]`, where potentially `i=0`. You can possibly index into `bins[-1]` in this case, and boom, your computer is up in smoke.

2) I recommend replacing this by two standard library calls. First use [`std::is_sorted`](https://en.cppreference.com/w/cpp/algorithm/is_sorted) to check they're in ascending order (you may need to manually set it to use the <= operator rather than the < operator as is used by default, although I think the < operator is what should be used here anyway). After checking they're in order, push the contents of `bins` into `bins_` by using `std::copy(bins.begin(), bins.end(), std::back_inserter(bins_));`. So this all should be replaced by like three lines of code.

Lastly, I think we are leaning towards calls to `openmc::fatal_error` rather than throwing exceptions now, but @paulromano will have to check me on that one. This is so we can eventually make a our own unified error handling system this way that interfaces with python/GPU perfectly, where exceptions don't work as cleanly.

Reply:
Yeah, for now I would recommend just calling `fatal_error`

Reply:
Thinking about this a little bit more, I don't see why we should need to enforce monotonicity at all. We can store an unordered map of (number of collisions) → (index in bins) that will allow us to get the correct index in O(1) time. You can look at CellFilter for an example of how this would work.

Reply:
So you'd recommend removing the ordered check entirely? I'm alright with that. I suppose it while it seems most natural to put them in increasing order there is no real reason to enforce that.

Reply:
Exactly, might as well give the user the flexibility to order them however they want since we can do O(1) lookups with an unordered map anyway